lib/deep_preloader.rb

```diff
@@ -40,23 +24,44 @@ def initialize(lock:)
       @worklist = {}
     end
 
-    def add_associations_from_spec(models, model_class, spec)
-      spec.association_specs.each do |association_name, child_spec|
-        association_reflection = model_class.reflect_on_association(association_name)
-        if association_reflection.nil?
-          raise ArgumentError.new("Preloading error: couldn't find association #{association_name} on model class #{model_class.name}")
+    def add_associations_from_spec(models, spec)
+      models = Array.wrap(models)
+
+      if spec.polymorphic?
```

Comment:
Took a cut at it: how does that look?

Reply:
Hopefully that's enough. It's still slightly confusing that polymorphic specs are used in both cases. In case we refer back to this later:

The polymorphic spec specifies the desired behavior for a given model type. Both cases have a single association that will result in different model types. If the reference contains the type (polymorphic association) then the polymorphic spec is eliminated while loading the children of each type. If the reference does not contain the type (STI) the children are all loaded in one call and the polymorphic spec is eliminated here.